lib/ddtrace/sampler.rb

```diff
@@ -141,10 +154,19 @@ def sample!(span)
         # If priority sampling has already been applied upstream, use that, otherwise...
         unless priority_assigned_upstream?(span)
           # Roll the dice and determine whether how we set the priority.
+          priority = priority_sample!(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+
           # NOTE: We'll want to leave `span.sampled = true` here; all spans for priority sampling must
           #       be sent to the agent. Otherwise metrics for traces will not be accurate, since the
           #       agent will have an incomplete dataset.
-          priority = priority_sample(span) ? Datadog::Ext::Priority::AUTO_KEEP : Datadog::Ext::Priority::AUTO_REJECT
+          #
+          #       We also ensure that the sampling rate is not propagated to the agent, to avoid erroneous
+          #       metric upscaling.
+          span.sampled = true
+          # TODO: do i have to unset SAMPLE_RATE_METRIC_KEY? or setting it to 1.0 is fine?
+          # TODO: a method to unset doesn't exist yet.
+          span.set_metric(SAMPLE_RATE_METRIC_KEY, 1.0) if span.get_metric(SAMPLE_RATE_METRIC_KEY)
```

Comment:
This should only be set on "RateSampler" a.k.a "client sampling".

For a sampler which sets `span.sampled = false` we need to set this metric to tell the agent how to upscale metrics for the traces that are not sent to it.

For "priority sampling" every span is sampled and sent to the agent, so we should not set this.

Reply:
Would this mean that sending `1.0` on `_sample_rate` for a sampler that always sets `span.sampled = true` is correct?

Reply:
This has been properly addressed: the tag is only set we are pre-sampling in memory and not sending all spans to the agent, as it has always been.